Initial review

titanous opened the pull request.

progrium committed:
initial commit with backend and maybe a working rpc server

progrium committed:
didn't mean to include this

progrium committed:
working version, see demo example

progrium committed:
whitespace

progrium committed:
whitespace

titanous committed:
gofmt

progrium committed:
Merge pull request #1 from flynn/gofmt

gofmt

titanous committed:
Merge branch 'master' into initial

titanous commented on .gitignore:
```diff
@@ -0,0 +1 @@
+discoverd/discoverd
```
I always add `*.test` to my Go gitignores, because `go test` leaves `package.test` binaries around.

progrium replied:
I haven't seen then when using `go test`, but clearly it doesn't hurt to add it.

titanous replied:
Is your `go version` `1.1.2`?

titanous commented on discover/backend_etcd.go:
```diff
@@ -0,0 +1,85 @@
+	"fmt"
+	"github.com/coreos/etcd/store"
+	"github.com/coreos/go-etcd/etcd"
+	"strings"
```
I've recently been using an import style that splits out non-stdlib imports for clarity:

``` go
import (
  "fmt"
  "strings"

  "github.com/coreos/etcd/store"
  "github.com/coreos/go-etcd/etcd"
)
```

`gofmt` handles sorting the two sections properly.

progrium replied:
Okay cool, sort of Pep8-ish.

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,194 @@
+	Port   string
+	Addr   string
+	Attrs  map[string]string
+	Online bool
```
How is the "array index" of the service indicated? (one example of usage is for leader election; master is 0, slaves are >0)

progrium replied:
I haven't included that yet because I haven't looked into whether the list of items returned from etcd is consistently ordered and by how. Though I've seen an example that relies on consistent ordering based on what seems to be index created, it's not documented, nor is that "index created" property exposed. The index returned for an entry is always the last index it was touched. 

But yes, probably some kind of index property would be on a service.

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,194 @@
+func NewClient() *DiscoverClient {
+	client, err := rpcplus.DialHTTP("tcp", "127.0.0.1:1111") // TODO: default, not hardcoded
+	if err != nil {
+		log.Fatal("dialing:", err)
```
This function should return `(*DiscoverClient, error)`, we don't want to call `exit` from library code.

progrium replied:
Good call.

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,194 @@
+	if err != nil {
+		log.Fatal("dialing:", err)
+	}
+	return &DiscoverClient{client: client, heartbeats: map[string]bool{}}
```
It's more idiomatic to use `make(map[string]bool)`.

progrium replied:
Rgr

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,194 @@
+		for update := range updates {
+			// TODO: apply filters
+			_, exists := s.services[update.Addr]
+			if !exists {
```
This can optionally be collapsed to one line:

``` go
if _, exists := s.services[update.Addr]; !exists {
```

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,194 @@
+	Name   string
+	Host   string
+	Port   string
+	Addr   string
```
Is there a reason to have `Host`/`Port` _and_ `Addr`? 99% of the `net` APIs will just want `Addr`.

progrium replied:
Right, but almost every other language separates them, so I was going for an API that would be more consistent in the long term, where you have them separate and Addr would be whatever canonical representation in that environment -- for example python: (host, ip) tuple.

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+type ServiceSet struct {
+	services  map[string]*Service
```
This map needs a `sync.RWMutex` protecting it. The map implementation is not thread-safe.

titanous replied:
I think all of these maps do, actually.

progrium replied:
Okay, I was wondering about that.

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+func (s *ServiceSet) OfflineAddrs() []string {
+	list := []string{}
```
Do a `make([]string, 0, len(s.service))` to allocate the whole slice in one go. (Ditto for the rest of the methods returning `[]string`)

progrium replied:
Rgr

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,194 @@
+				}, &heartbeated)
+				time.Sleep(HeartbeatIntervalSecs * time.Second) // TODO: add jitter
+			}
+		}(name, args.Addr)
```
You shouldn't need to pass these args in explicitly. It _is_ necessary in some `for` loops, but in this case it should close over the variables properly and escape analysis will kick in.

progrium replied:
Okay, I wanted to be sure because I ran into that case before.

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,194 @@
+		}(name, args.Addr)
+		return nil
+	} else {
+		return errors.New("register failed")
```
Idiomatic style would be to check right away:

``` go
if !success {
  return errors.New("register failed")
}
```

and then outdent the success path below the check.

progrium replied:
Rgr

titanous replied:
Also, in packages we prefix errors with the package name, so this should be `"discover: register failed"`.

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,194 @@
+	return c.RegisterWithHost(name, pickMostPublicIp(), port, attributes)
+}
+
+func (c *DiscoverClient) RegisterWithHost(name string, host string, port string, attributes map[string]string) error {
```
Since all of the `net` APIs deal with `addr`, I'd use an `addr string` instead of `host`/`port`. Also, you can just write `string` once:

``` go
func (c *DiscoverClient) RegisterWithHost(name, addr string, attributes map[string]string) error { 
```

progrium replied:
See previous comment on combining host/port. But the idiomatic API would be the main Register function anyway.

titanous commented on discover/server.go:
```diff
@@ -0,0 +1,97 @@
+	}
+}
+
+func ServeForever(server *DiscoverAgent) {
```
The idiomatic name for this function is `ListenAndServe` and it should return an error if `Listen` fails.

progrium replied:
Rgr. I knew there was something better. Borrowed from Python here.

titanous commented on discover/server.go:
```diff
@@ -0,0 +1,97 @@
+
+func ServeForever(server *DiscoverAgent) {
+	rpcplus.Register(server)
+	rpcplus.HandleHTTP()
```
Is there a reason to use the HTTP variant of the server? It's less overhead to use the vanilla version (`Accept`).

progrium replied:
I was using the debug endpoint. HTTP might nice eventually for other reasons, but agree it's not worth it now.

titanous commented on discover/server.go:
```diff
@@ -0,0 +1,97 @@
+}
+
+func (s *DiscoverAgent) Register(args *Args, success *bool) error {
+	err := s.Backend.Register(args.Name, args.Addr, map[string]string{}) // TODO: attrs!
```
For reference you can typically just use `nil` instead of an empty map.

progrium replied:
Rgr

titanous commented on discover/server.go:
```diff
@@ -0,0 +1,97 @@
+	} else {
+		*success = true
+	}
+	return nil
```
I think we want to actually return `err` here. You can use `struct{}` as an empty return argument.

progrium replied:
What do you mean? Multiple arguments? Or do you mean the RPC should use an error object for response?

titanous replied:
This function should look something like:

``` go
func (s *DiscoverAgent) Register(args *Args, ret *struct{}) error {
  return s.Backend.Register(args.Name, args.Addr, nil)
}
```

titanous commented on discover/server.go:
```diff
@@ -0,0 +1,97 @@
+	Subscribe(string) (chan *ServiceUpdate, error)
+	Register(string, string, map[string]string) error
+	Unregister(string, string) error
+	Heartbeat(string, string) error
```
I like naming my arguments in interface definitions.

progrium replied:
Rgr, me too. Wasn't sure what was more idiomatic.

titanous commented on discover/backend_etcd.go:
```diff
@@ -0,0 +1,85 @@
+			Attrs:  map[string]string{},
+		}
+	}
+	return nil
```
It might be clearer to move this `return` up into a `default` clause of the `case` statement.

progrium replied:
Rgr

titanous commented on discover/backend_etcd.go:
```diff
@@ -0,0 +1,85 @@
+	if err != nil {
+		return err
+	}
+	return nil
```
Just `return err` here.

progrium replied:
Yep, silly me.

titanous commented on discover/backend_etcd.go:
```diff
@@ -0,0 +1,85 @@
+func (b *EtcdBackend) responseToUpdate(resp *store.Response) *ServiceUpdate {
+	serviceName := strings.SplitN(resp.Key, "/", 4)[2]
+	switch {
+	case ("SET" == resp.Action && resp.NewKey) || "GET" == resp.Action:
```
I'd prefer this as a `if` statement, unless we drop the `resp.NewKey` condition. What is the reason to drop SETs on existing keys?

progrium replied:
SETs on existing are heartbeats with no changes, which we don't care about in this abstraction. We just want to know when it changes. So make this an `if` statement?

titanous replied:
Yep, `if`.

titanous commented on discover/backend_etcd.go:
```diff
@@ -0,0 +1,85 @@
+			Name:   serviceName,
+			Addr:   resp.PrevValue,
+			Online: false,
+			Attrs:  map[string]string{},
```
No need to include `Online` or `Attrs`, they default to their zero values (`false` and `map[string]string(nil)`) respectively.

progrium replied:
Rgr

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,194 @@
+	}, updates)
+	set := &ServiceSet{
+		services: map[string]*Service{},
+		filters:  map[string]string{},
```
Use `make` for these.

titanous commented on examples/client.go:
```diff
@@ -0,0 +1,34 @@
+	"log"
+)
+
+func main() {
```
Hmm, do we even want this example? I don't want to encourage anyone to re-implement `discover/client.go` in their Go apps.

progrium replied:
You're right. It was a debugging tool.

titanous commented on examples/demo.go:
```diff
@@ -0,0 +1,47 @@
+
+	exit := make(chan os.Signal, 1)
+	signal.Notify(exit, os.Interrupt)
+	signal.Notify(exit, syscall.SIGTERM)
```
`signal.Notify` is variadic:

``` go
signal.Notify(exit, os.Interrupt, syscall.SIGTERM)
```

progrium replied:
Rgr

titanous commented on examples/demo.go:
```diff
@@ -0,0 +1,47 @@
+	flag.Parse()
+	name := flag.Arg(0)
+	port := flag.Arg(1)
+	host := flag.Arg(2)
```
No need for `flag`, you can just use `os.Args`.

progrium replied:
Rgr

progrium replied:
Ended up going back to flag.Arg because os.Args requires checking the index for optional args, where flag.Arg gives an empty string. Slightly simpler.

titanous commented on examples/demo.go:
```diff
@@ -0,0 +1,47 @@
+
+	set := client.Services(name)
+	for {
+		fmt.Printf("%s\n", strings.Join(set.OnlineAddrs(), ", "))
```
Use `fmt.Println`.

progrium replied:
Heh, yep.

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,194 @@
+	var ip string
+	for _, addr := range addrs {
+		ip = strings.SplitN(addr.String(), "/", 2)[0]
+		if !strings.Contains(ip, "::") && ip != "127.0.0.1" {
```
There are a bunch of loopback addresses, this will probably be problematic.

progrium replied:
Oh? Like what? If so, aren't they common?

titanous replied:
Yeah, just looked it up. `127.0.0.0/8` and `::1`.

titanous commented on discover/backend_etcd.go:
```diff
@@ -0,0 +1,85 @@
+	for _, response := range responses {
+		update := b.responseToUpdate(response)
+		if update != nil {
+			ch <- update
```
It looks like we can deadlock here if there are more than 10 updates?

titanous replied:
Also, doesn't this need to be done _after_ we subscribe to state changes for consistency?

progrium replied:
For first point, yes. I realized it's probably better to take a chan as argument. Though that just moves the problem elsewhere, but maybe we can at least make sure there is a receiver before we call Subscribe. 

You're right about consistency, but with etcd we can just include the index of that last update received and we shouldn't miss anything. If we can do that it reads better this way at least.

titanous commented on discover/backend_etcd.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+func (b *EtcdBackend) getStateChanges(name string) chan *store.Response {
+	watch := make(chan *store.Response, 10)
```
Is there a reason to buffer this channel? If so, please add a comment explaining the significance of the buffer.

progrium replied:
There isn't since we can usually be sure there is a receiver, but from other code I've read it seemed like good practice to buffer anyway. What is the downside?

titanous replied:
Typically you don't add a buffer unless there's a good reason, as it causes the semantics of channels change. As pointed out above, it becomes easy to create code that can deadlock under some conditions but not others.

titanous commented on discover/backend_etcd.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+func (b *EtcdBackend) responseToUpdate(resp *store.Response) *ServiceUpdate {
+	serviceName := strings.SplitN(resp.Key, "/", 4)[2]
```
Is it possible for resp.Key to _ever_ have less than 3 segments? If so, this will cause a runtime panic.

progrium replied:
Offensive programming at its worst. Thanks.

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,194 @@
+
+// Still not sure about this API, but it's a start
+func (s *ServiceSet) Listen(ch chan *ServiceUpdate) {
+	s.listeners[ch] = struct{}{}
```
I think we need a way to remove listeners too.

progrium replied:
Absolutely, just wasn't sure on this API to begin with, so it was left unfinished. Is it reasonable? Listen and Unlisten?

titanous replied:
Yep, or `Subscribe` and `Unsubscribe`.

progrium replied:
Ah that works. Before `Services` was `Subscribe` so it would have been confusing, but it would work now.

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,194 @@
+
+type DiscoverClient struct {
+	client     *rpcplus.Client
+	heartbeats map[string]bool
```
This map needs a mutex.

progrium replied:
Rgr

progrium commented:
This is awesome. I probably won't get back to this until next week though.

georgethomas111 committed:
Adding some review changes. Build successful, untested.

georgethomas111 committed:
1. Added Changes as per review comments.
2. Added a test case for client.go. (Need to think of a
method to start the etcd/backend server from the testcase
Currently etcd needs to be started manually)

georgethomas111 committed:
Corrected the import for my version

georgethomas111 committed:
Removed invalid comments

titanous committed:
Add LICENSE

progrium committed:
cleaning up george's contributions, touching remaining items from jonathan's review. overall, better code, more testing

progrium committed:
removed bad mutex. made sure demo still worked

progrium committed:
updating imports now that my fork is gone

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,206 @@
+}
+
+func (s *ServiceSet) Online() []*Service {
+	s.serMutex.Lock()
```
Add a `defer s.serMutex.Unlock()` here (and for the rest of these functions).

progrium replied:
rgr

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,206 @@
+	listeners map[chan *ServiceUpdate]struct{}
+	serMutex  sync.RWMutex
+	filMutex  sync.RWMutex
+	lisMutex  sync.RWMutex
```
I don't see you calling `RLock` at all so these can probably just be `sync.Mutex`? (Unless you switch to calling `RLock`)

progrium replied:
Are you opposed to using a regular mutex?

titanous replied:
Nope, totally fine with me.

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,206 @@
+}
+
+func (s *ServiceSet) Filter(attrs map[string]string) {
+	s.filters = attrs
```
Lock `s.filMutex` here?

progrium replied:
Yeah, I just haven't really implemented filters yet so I didn't touch this.

progrium replied:
Actually, because this is an assignment replacing filters, does it really need a mutex? It's not an operation on a map...

titanous replied:
Yeah, swapping a pointer isn't guaranteed atomic unless you use `sync/atomic`.

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,206 @@
+}
+
+func (c *DiscoverClient) Services(name string) *ServiceSet {
+	updates := make(chan *ServiceUpdate, 10)
```
What is this buffer for?

progrium replied:
Hmm, I thought I took that out.

titanous commented on discover/client.go:
```diff
@@ -0,0 +1,206 @@
+				Name: name,
+				Addr: args.Addr,
+			}, &heartbeated)
+			time.Sleep(HeartbeatIntervalSecs * time.Second) // TODO: add jitter
```
Can we eliminate one of these calls to sleep?

progrium replied:
Obviously putting the second one at the beginning of the loop would eliminate the need for the first one, but I feel like there was good reason for this. I just can't remember it now, so maybe not.

progrium committed:
putting unlock under defer

progrium committed:
simplifying heartbeat sleeps

progrium committed:
removing unnecessary channel buffer

progrium committed:
changing to regular Mutex since not using RLock of RWMutex

titanous commented:
LGTM. Have you tested with `go test -race`?
